pppd/sys-linux.c

```diff
@@ -1469,9 +1476,131 @@ get_ppp_stats(int u, struct pppd_stats *stats)
     stats->bytes_out = req.stats.p.ppp_obytes;
     stats->pkts_in = req.stats.p.ppp_ipackets;
     stats->pkts_out = req.stats.p.ppp_opackets;
+
+    if (stats->bytes_in < previbytes)
+	++iwraps;
+    if (stats->bytes_out < prevobytes)
+	++owraps;
+
+    previbytes = stats->bytes_in;
+    prevobytes = stats->bytes_out;
+
+    stats->bytes_in += (u_int64_t)iwraps << 32;
+    stats->bytes_out += (u_int64_t)owraps << 32;
+
+    return 1;
+}
+
+/********************************************************************
+ * get_ppp_stats_sysfs - return statistics for the link, using the files in sysfs,
+ * this provides native 64-bit counters.
+ */
+static int
+get_ppp_stats_sysfs(int u, struct pppd_stats *stats)
+{
+    char fname[PATH_MAX+1];
+    char buf[21], *err; /* 2^64 < 10^20 */
+    int blen, fd, rlen;
+    unsigned long long val;
+
+    struct {
+	const char* fname;
+	void* ptr;
+	unsigned size;
+    } slist[] = {
+#define statfield(fn, field)	{ .fname = #fn, .ptr = &stats->field, .size = sizeof(stats->field) }
+	statfield(rx_bytes, bytes_in),
+	statfield(tx_bytes, bytes_out),
+	statfield(rx_packets, pkts_in),
+	statfield(tx_packets, pkts_out),
+#undef statfield
+    };
+
+    blen = snprintf(fname, sizeof(fname), "/sys/class/net/%s/statistics/", ifname);
+    if (blen >= sizeof(fname))
+	return 0; /* ifname max 15, so this should be impossible */
+
+    for (int i = 0; i < sizeof(slist) / sizeof(*slist); ++i) {
+	if (snprintf(fname + blen, sizeof(fname) - blen, "%s", slist[i].fname) >= sizeof(fname) - blen) {
+	    fname[blen] = 0;
+	    error("sysfs stats: filename %s/%s overflowed PATH_MAX", fname, slist[i].fname);
+	    return 0;
+	}
+
+	fd = open(fname, O_RDONLY);
+	if (fd < 0) {
+	    error("%s: %m", fname);
+	    return 0;
+	}
+
+	rlen = read(fd, buf, sizeof(buf) - 1);
+	close(fd);
+	if (rlen < 0) {
+	    error("%s: %m", fname);
+	    return 0;
+	}
+	/* trim trailing \n if present */
+	while (rlen > 0 && buf[rlen-1] == '\n')
+	    rlen--;
+	buf[rlen] = 0;
+
+	val = strtoull(buf, &err, 10);
+	if (err && *err) {
```

Comment:
This is incorrect error checking if `strtoull` failed. `strto*` has lot of traps, look into [documentation](https://man7.org/linux/man-pages/man3/strtol.3.html). You have to check also for `errno` and validity of the first byte in `buf`.

Reply:
```
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <errno.h>

int main(int, char **argv)
{
	unsigned long long r;
	char *endptr;

	while (*++argv) {
		r = strtol(*argv, &endptr, 0);
		printf("i=%p\"%s\" r=%llu, errno=%d, endptr=%p\"%s\"(%lu)\n",
				*argv, *argv, r, errno, endptr, endptr, strlen(endptr));
	}
	return 0;
}
```

Now this is interesting (ERANGE==34):

```
jkroon@plastiekpoot ~/src/test $ ./strtol 01a "" "-1" "18446744073709551616" 0 1
i=0x7ffdcd69bfaf"01a" r=1, errno=0, endptr=0x7ffdcd69bfb1"a"(1)
i=0x7ffdcd69bfb3"" r=0, errno=0, endptr=0x7ffdcd69bfb3""(0)
i=0x7ffdcd69bfb4"-1" r=18446744073709551615, errno=0, endptr=0x7ffdcd69bfb6""(0)
i=0x7ffdcd69bfb7"18446744073709551616" r=18446744073709551615, errno=34, endptr=0x7ffdcd69bfcb""(0)
i=0x7ffdcd69bfcc"0" r=0, errno=34, endptr=0x7ffdcd69bfcd""(0)
i=0x7ffdcd69bfce"1" r=1, errno=34, endptr=0x7ffdcd69bfcf""(0)
```

I'd expect -1 to also set errno=ERANGE (and return 0, but ULONG_MAX is in line with documentation).  I'd expect strtoll (if it's going to use errno for returning errors) to zero it upon entry.  "lots of traps" indeed.  Hopeing it's covered now.

Reply:
Yes, `strtou...` functions do not signal underflow. This is really a trap. Anyway, you do not need to test `err` pointer for non-NULL. After successful call (when errno == 0) is `err` set, so you can dereference `*err`.

Reply:
Dropped the check for err being non-null.